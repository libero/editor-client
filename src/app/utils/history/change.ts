import { Manuscript } from 'app/types/manuscript';

export abstract class Change {
  abstract applyChange(manuscript: Manuscript): Manuscript;
  abstract rollbackChange(manuscript: Manuscript): Manuscript;
  abstract get isEmpty(): boolean;

  private _timestamp: number;
  constructor() {
    this._timestamp = Date.now();
  }

  get timestamp(): number {
    return this._timestamp;
  }
}

export class BatchChange extends Change {
  private changes: Change[];
  constructor(changes: Change[] = []) {
    super();
    this.changes = changes.filter((change) => !change.isEmpty);
  }

  get isEmpty(): boolean {
    return !Array.isArray(this.changes) || this.changes.length === 0;
  }

  applyChange(manuscript: Manuscript): Manuscript {
    return this.changes.reduce(
      (acc: Manuscript, change: Change) => {
        return change.applyChange(acc);
      },
      { ...manuscript }
    );
  }

  rollbackChange(manuscript: Manuscript): Manuscript {
    return this.changes.reduceRight(
      (acc: Manuscript, change: Change) => {
        return change.rollbackChange(acc);
      },
      { ...manuscript }
    );
  }
}
