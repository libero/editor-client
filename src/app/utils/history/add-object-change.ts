import { get, set } from 'lodash';

import { Manuscript } from 'app/types/manuscript';
import { Change } from 'app/utils/history/change';
import { cloneManuscript } from 'app/utils/state.utils';
import { JSONObject } from 'app/types/utility.types';
import { manuscriptEntityToJson } from 'app/utils/changes.utils';

export class AddObjectChange<T> extends Change {
  constructor(private path: string, private object: T, private idField: string) {
    super();
  }

  get isEmpty(): boolean {
    return false;
  }

  applyChange(manuscript: Manuscript): Manuscript {
    const originalSection = get(manuscript, this.path);

    if (!Array.isArray(originalSection)) {
      throw new TypeError('Trying to make AddObject change on a non-array section');
    }

    return set(cloneManuscript(manuscript), this.path, [...originalSection, this.object]);
  }

  rollbackChange(manuscript: Manuscript): Manuscript {
    const originalSection = get(manuscript, this.path);

    if (!Array.isArray(originalSection)) {
      throw new TypeError('Trying to rollback AddObject change on a non-array section');
    }
    const updatedSection = originalSection.filter((item) => item[this.idField] !== this.object[this.idField]);
    return set(cloneManuscript(manuscript), this.path, updatedSection);
  }

  toJSON(): JSONObject {
    return {
      type: 'add-object',
      timestamp: this.timestamp,
      path: this.path,
      idField: this.idField,
      object: manuscriptEntityToJson<T>(this.object)
    };
  }
}
