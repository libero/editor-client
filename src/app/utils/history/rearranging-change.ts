import { get, set } from 'lodash';

import { Manuscript } from '../../types/manuscript';
import { Change } from './change';
import { cloneManuscript } from '../state.utils';
import { JSONObject } from '../../types/utility.types';

export class RearrangingChange extends Change {
  public static createFromListRearrange<T>(path: string, oldOrder: Array<T>, newOrder: Array<T>): RearrangingChange {
    const order = newOrder.map((item) => oldOrder.indexOf(item));
    return new RearrangingChange(path, order);
  }

  public static createFromItemMoved<T>(
    path: string,
    oldIndex: number,
    newIndex: number,
    collection: Array<T>
  ): RearrangingChange {
    const order = new Array(collection.length).fill(undefined).map((_, index) => index);
    order.splice(oldIndex, 1);
    order.splice(newIndex, 0, oldIndex);
    return new RearrangingChange(path, order);
  }

  public static fromJSON<T>(data: JSONObject): RearrangingChange {
    const change = new RearrangingChange(data.path as string, data.order as Array<number>);
    change._timestamp = data.timestamp as number;
    return change;
  }

  private constructor(private path: string, private order: number[]) {
    super();
  }

  get isEmpty(): boolean {
    return !this.order.some((el, index) => el !== index);
  }

  isPathAffected(pathPattern: RegExp): boolean {
    return pathPattern.test(this.path);
  }

  applyChange(manuscript: Manuscript): Manuscript {
    return set(cloneManuscript(manuscript), this.path, this.applyOrder(get(manuscript, this.path), this.order));
  }

  rollbackChange(manuscript: Manuscript): Manuscript {
    const rollbackOrder = this.reverseOrder(this.order);
    return set({ ...manuscript }, this.path, this.applyOrder(get(manuscript, this.path), rollbackOrder));
  }

  toJSON(): JSONObject {
    return {
      type: 'rearranging',
      timestamp: this.timestamp,
      path: this.path,
      order: this.order
    };
  }

  private applyOrder<T>(collection: Array<T>, order: number[]): Array<T> {
    return order.map((index) => collection[index]);
  }

  private reverseOrder(order: number[]): number[] {
    return order.reduce((acc, oldIndex, newIndex) => {
      acc[oldIndex] = newIndex;
      return acc;
    }, new Array<number>(order.length));
  }
}
