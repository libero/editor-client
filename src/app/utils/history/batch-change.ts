import { Manuscript } from 'app/types/manuscript';
import { JSONObject } from 'app/types/utility.types';
import { deserializeChanges } from 'app/utils/changes.utils';
import { Change } from 'app/utils/history/change';

export class BatchChange extends Change {
  static fromJSON(data: JSONObject): BatchChange {
    const change = new BatchChange(deserializeChanges(data.changes as JSONObject[]));
    change._timestamp = data.timestamp as number;
    return change;
  }

  private changes: Change[];
  constructor(changes: Change[] = []) {
    super();
    this.changes = changes.filter((change) => !change.isEmpty);
  }

  get isEmpty(): boolean {
    return !Array.isArray(this.changes) || this.changes.length === 0;
  }

  applyChange(manuscript: Manuscript): Manuscript {
    return this.changes.reduce((acc: Manuscript, change: Change) => {
      return change.applyChange(acc);
    }, manuscript);
  }

  rollbackChange(manuscript: Manuscript): Manuscript {
    return this.changes.reduceRight((acc: Manuscript, change: Change) => {
      return change.rollbackChange(acc);
    }, manuscript);
  }

  toJSON(): JSONObject {
    return {
      type: 'batch',
      changes: this.changes.map((change) => change.toJSON()),
      timestamp: this.timestamp
    };
  }
}
