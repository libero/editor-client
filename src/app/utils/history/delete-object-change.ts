import { Manuscript } from 'app/types/manuscript';
import { get, set } from 'lodash';
import { Change } from 'app/utils/history/change';
import { cloneManuscript } from 'app/utils/state.utils';

export class DeleteObjectChange<T> implements Change {
  private removedIndex: number;
  constructor(private path: string, private object: T, private idField: string) {}

  get isEmpty(): boolean {
    return false;
  }

  applyChange(manuscript: Manuscript): Manuscript {
    const originalSection = get(manuscript, this.path);

    if (!Array.isArray(originalSection)) {
      throw new TypeError('Trying to make AddObject change on a non-array section');
    }
    this.removedIndex = originalSection.findIndex((item) => item[this.idField] === this.object[this.idField]);
    const updatedSection = [...originalSection];
    updatedSection.splice(this.removedIndex, 1);
    return set(cloneManuscript(manuscript), this.path, updatedSection);
  }

  rollbackChange(manuscript: Manuscript): Manuscript {
    const originalSection = get(manuscript, this.path);

    if (!Array.isArray(originalSection)) {
      throw new TypeError('Trying to rollback AddObject change on a non-array section');
    }
    const updatedSection = originalSection.filter((item) => item[this.idField] === this.object[this.idField]);
    return set(cloneManuscript(manuscript), this.path, updatedSection);
  }
}
