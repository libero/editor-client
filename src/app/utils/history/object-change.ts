import * as deepDiff from 'deep-diff';
import { Manuscript } from 'app/types/manuscript';
import { cloneDeep, get, set } from 'lodash';
import { Change } from 'app/utils/history/change';
import { cloneManuscript } from 'app/utils/state.utils';

export class ObjectChange<T> implements Change {
  private differences: deepDiff.Diff<T, T>[];

  constructor(private path: string, oldObject: T, newObject: T) {
    this.differences = deepDiff.diff(oldObject, newObject);
  }

  get isEmpty(): boolean {
    return !Array.isArray(this.differences) || this.differences.length === 0;
  }

  applyChange(manuscript: Manuscript): Manuscript {
    const originalSection = get(manuscript, this.path);

    const updatedSection = this.differences.reduce((acc: T, diff) => {
      const newAcc = cloneDeep(acc);
      deepDiff.applyChange(newAcc, acc, diff);
      return newAcc;
    }, cloneDeep(originalSection));

    return set(cloneManuscript(manuscript), this.path, updatedSection);
  }

  rollbackChange(manuscript: Manuscript): Manuscript {
    const originalSection = get(manuscript, this.path);

    const updatedSection = this.differences.reduce((acc: T, diff) => {
      const newAcc = cloneDeep(acc);
      deepDiff.revertChange(newAcc, acc, diff);
      return newAcc;
    }, originalSection);

    return set(cloneManuscript(manuscript), this.path, updatedSection);
  }
}
