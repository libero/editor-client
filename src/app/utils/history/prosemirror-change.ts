import { Transaction } from 'prosemirror-state';
import { Manuscript } from 'app/types/manuscript';
import { get, set } from 'lodash';
import { Change } from 'app/utils/history/change';
import { cloneManuscript } from 'app/utils/state.utils';

export class ProsemirrorChange implements Change {
  constructor(private path: string, private transaction: Transaction) {}

  get isEmpty(): boolean {
    return !this.transaction.docChanged;
  }

  applyChange(manuscript: Manuscript): Manuscript {
    const editorState = get(manuscript, this.path);
    return set(cloneManuscript(manuscript), this.path, editorState.apply(this.transaction));
  }

  rollbackChange(manuscript: Manuscript): Manuscript {
    const invertedSteps = this.transaction.steps.map((step, index) => {
      return step.invert(this.transaction.docs[index]);
    });

    const editorState = get(manuscript, this.path);
    const rollbackTransaction = editorState.tr;
    invertedSteps.reverse().forEach((step) => rollbackTransaction.maybeStep(step));

    return set(cloneManuscript(manuscript), this.path, editorState.apply(rollbackTransaction));
  }
}
