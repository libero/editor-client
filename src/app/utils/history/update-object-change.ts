import * as deepDiff from 'deep-diff';
import { cloneDeepWith, get, set } from 'lodash';
import { EditorState, Transaction } from 'prosemirror-state';

import { Manuscript } from 'app/types/manuscript';
import { Change } from 'app/utils/history/change';
import { cloneManuscript } from 'app/utils/state.utils';
import { JSONObject } from 'app/types/utility.types';
import { ProsemirrorChange } from 'app/utils/history/prosemirror-change';
import { BatchChange } from 'app/utils/history/batch-change';

export class UpdateObjectChange<T> extends Change {
  public static createFromTwoObjects<T>(path: string, oldObject: T, newObject: T): Change {
    const editorStateProps = [];

    const pojoDifferences = deepDiff.diff(
      oldObject,
      newObject,
      // if prefilter function returns true diff stops drilling down
      (path: Array<string | number>, key: string | number) => {
        const propPath = [...path, key].join('.');
        if (get(oldObject, propPath) instanceof EditorState) {
          editorStateProps.push(propPath);
          return true;
        }
        return false;
      }
    );

    const objectChange = new UpdateObjectChange(path, pojoDifferences);
    const prosemirrorChanges = editorStateProps
      .map((propPath: string) => {
        const transaction = UpdateObjectChange.getEditorStatesDiff(get(oldObject, propPath), get(newObject, propPath));
        if (transaction.docChanged) {
          return new ProsemirrorChange(`${path}.${propPath}`, transaction);
        }
        return undefined;
      })
      .filter(Boolean);

    return new BatchChange([objectChange, ...prosemirrorChanges]);
  }

  public static fromJSON<T>(data: JSONObject): UpdateObjectChange<T> {
    const change = new UpdateObjectChange(data.path as string, (data.differences as unknown) as deepDiff.Diff<T, T>[]);
    change._timestamp = data.timestamp as number;
    return change;
  }

  private constructor(private path: string, private differences: deepDiff.Diff<T, T>[] | undefined) {
    super();
  }

  get isEmpty(): boolean {
    return !Array.isArray(this.differences) || this.differences.length === 0;
  }

  applyChange(manuscript: Manuscript): Manuscript {
    const originalSection = get(manuscript, this.path);
    const updatedSection = this.differences.reduce((acc: T, diff) => {
      const newAcc = this.cloneWithoutEditorState(acc);
      deepDiff.applyChange(newAcc, acc, diff);
      return newAcc;
    }, originalSection);

    return set(cloneManuscript(manuscript), this.path, updatedSection);
  }

  rollbackChange(manuscript: Manuscript): Manuscript {
    const originalSection = get(manuscript, this.path);

    const updatedSection = this.differences.reduce((acc: T, diff) => {
      const newAcc = this.cloneWithoutEditorState(acc);
      deepDiff.revertChange(newAcc, acc, diff);
      return newAcc;
    }, originalSection);

    return set(cloneManuscript(manuscript), this.path, updatedSection);
  }

  toJSON(): JSONObject {
    return {
      type: 'update-object',
      timestamp: this.timestamp,
      path: this.path,
      differences: (this.differences as unknown) as JSONObject
    };
  }

  private cloneWithoutEditorState(object: T): T {
    const cloneCustomizer = (value): EditorState | undefined => (value instanceof EditorState ? value : undefined);
    return cloneDeepWith(object, cloneCustomizer);
  }

  private static getEditorStatesDiff(prevState: EditorState, nextState: EditorState): Transaction {
    const transaction = prevState.tr;
    const start = nextState.doc.content.findDiffStart(prevState.doc.content);
    if (start !== null) {
      let { a: endA, b: endB } = nextState.doc.content.findDiffEnd(get(prevState, 'doc.content'));
      const overlap = start - Math.min(endA, endB);
      if (overlap > 0) {
        endA += overlap;
        endB += overlap;
      }
      transaction.replace(start, endB, nextState.doc.slice(start, endA));
    }
    return transaction;
  }
}
