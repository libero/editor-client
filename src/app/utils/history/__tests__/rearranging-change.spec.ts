import { get } from 'lodash';

import { givenState } from 'app/test-utils/reducer-test-helpers';
import { RearrangingChange } from 'app/utils/history/rearranging-change';
import { Keyword } from 'app/models/keyword';

describe('RearrangingChange', () => {
  const manuscript = givenState({}).data.present;
  beforeAll(() => {
    manuscript.keywordGroups['kwdGroup'] = {
      title: 'group',
      keywords: [new Keyword(), new Keyword(), new Keyword()],
      newKeyword: new Keyword()
    };
  });

  it('should apply rearranging change', () => {
    const path = 'keywordGroups.kwdGroup.keywords';
    const change = RearrangingChange.createFromItemMoved(path, 2, 0, get(manuscript, path));
    const updatedManuscript = change.applyChange(manuscript);
    expect(get(updatedManuscript, path)).toEqual([
      get(manuscript, path)[2],
      get(manuscript, path)[0],
      get(manuscript, path)[1]
    ]);
  });

  it('should apply rearranging change from two lists', () => {
    const path = 'keywordGroups.kwdGroup.keywords';
    const change = RearrangingChange.createFromListRearrange(path, get(manuscript, path), [
      get(manuscript, path)[2],
      get(manuscript, path)[0],
      get(manuscript, path)[1]
    ]);

    const updatedManuscript = change.applyChange(manuscript);
    expect(get(updatedManuscript, path)).toEqual([
      get(manuscript, path)[2],
      get(manuscript, path)[0],
      get(manuscript, path)[1]
    ]);
  });

  it('should revert prosemirror change', () => {
    const path = 'keywordGroups.kwdGroup.keywords';
    const change = RearrangingChange.createFromItemMoved(path, 2, 0, get(manuscript, path));
    const updatedManuscript = change.applyChange(manuscript);

    expect(change.rollbackChange(updatedManuscript)).toEqual(manuscript);
  });

  it('should indicate empty change', () => {
    const path = 'keywordGroups.kwdGroup.keywords';
    const change = RearrangingChange.createFromItemMoved(path, 2, 2, get(manuscript, path));
    expect(change.isEmpty).toBeTruthy();
  });

  it('should serialize to JSON', () => {
    const path = 'keywordGroups.kwdGroup.keywords';
    const change = RearrangingChange.createFromItemMoved(path, 2, 0, get(manuscript, path));

    expect(change.toJSON()).toEqual({
      path: 'keywordGroups.kwdGroup.keywords',
      timestamp: expect.any(Number),
      type: 'rearranging',
      order: [2, 0, 1]
    });
  });

  it('should serialize to JSON', () => {
    const JSONObject = {
      path: 'keywordGroups.kwdGroup.keywords',
      timestamp: 1610979099826,
      order: [2, 0, 1],
      type: 'prosemirror'
    };

    const change = RearrangingChange.fromJSON(JSONObject);

    expect(change).toMatchSnapshot();
    expect(change).toBeInstanceOf(RearrangingChange);
  });
});
