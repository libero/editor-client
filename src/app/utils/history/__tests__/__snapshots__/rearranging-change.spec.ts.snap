// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RearrangingChange should serialize to JSON 1`] = `
Object {
  "order": Array [
    2,
    0,
    1,
  ],
  "path": "keywordGroups.kwdGroup.keywords",
  "timestamp": 1610979099826,
  "type": "rearranging",
}
`;
